### PR TITLE
fix(hooks): trunk-protection defense in depth — closes #64 #65 #68

### LINKED ISSUE #59: bug(dispatcher): resume can re-dispatch an already-completed session, causing wrapper to hang indefinitely on streaming socket
## Summary

When the dispatcher re-dispatches a dev task in `resume` mode against a session that has already completed normally (i.e. the prior `claude` invocation exited 0 with `stop_reason=end_turn`, `terminal_reason=completed`), the `claude --resume` call connects to the streaming endpoint and never returns — the wrapper hangs indefinitely.

## What I observed

A consumer project's dispatcher dispatched a dev task for an issue. The first wrapper ran the `claude` agent for ~6 hours, then exited cleanly (exit code 0) with the agent JSON output reporting `stop_reason: "end_turn"`, `terminal_reason: "completed"`. The wrapper transitioned the issue label correctly.

Later (≈2 hours after) a fresh dispatcher cycle ran. Because the issue label was still `autonomous` (this is normal — the issue is not closed; it's awaiting human review), the dispatcher matched the issue and dispatched again in `resume` mode against the same session-id. A new wrapper + new `claude` process started.

The new `claude` process:
- Connected to the Anthropic streaming endpoint
- Stayed connected for **8+ hours** with no new agent turns
- Used **~5 minutes total CPU over 8 hours** (sub-1% utilization)
- Read **~430 bytes / 10 seconds** on the streaming socket — purely SSE keepalive ping traffic, no business tokens
- `wchan = ep_pol` (epoll_wait), no progress

It only stopped when I sent SIGKILL.

## Hypothesis on the root cause

A `claude --resume <session-id>` against a session whose final turn already ended with `end_turn` has nothing to do, but the streaming endpoint will still happily open the SSE connection and keep it alive with periodic comments. The CLI client (with `--output-format json`) is waiting for a stop event that the server has no reason to emit, because the prior turn is the terminal one. Result: indefinite hang.

I have not confirmed this is exactly what the streaming endpoint does — but the symptoms (open SSE connection, keepalive bytes, zero output, indefinite epoll_wait) are consistent with it.

## Suggested fixes (any one of these would have prevented this)

1. **Dispatcher-side guard**: before issuing a `resume`, fetch the session's terminal state. If it's already `terminal_reason=completed` / `stop_reason=end_turn`, skip the dispatch entirely — there is nothing to resume. The agent already finished its job.

2. **Wrapper-side wall-clock timeout**: wrap the `run_agent` call in `timeout <N>h` so a hung resume cannot eat indefinite wall-clock time. See companion issue (filed separately) on the missing timeout.

3. **CLI-side fix** (out of scope for this repo, but worth flagging upstream to claude CLI): if the model returns no tokens for >5 minutes after a resume, abort with a non-zero exit code instead of waiting on SSE forever.

(1) is the most surgical and avoids the duplicate-work loop in the first place. (2) is a defense-in-depth net for any other "claude hangs" scenario.

## How this differs from #55 (closed)

#55 fixed the case where two wrappers run concurrently for the same issue. The fix in 66e2d0a kills the stale wrapper before spawning a new one. That's correct, but it doesn't address the case here:

- In #55: the prior dev session was still running / had crashed mid-turn → resume is a legitimate retry.
- Here: the prior dev session **completed normally**. There is nothing left to resume; the new wrapper is a redundant dispatch on a finished session, and the redundant dispatch is the thing that hangs.

So even with #55's fix in place, the dispatcher will still hang the wrapper if it dispatches `resume` against a `completed` session.

## Reproduction notes

I do not have a minimal repro on hand because reproducing it cleanly requires the full Claude CLI / Anthropic streaming endpoint flow. The above is from a real incident I diagnosed by attaching to the live process tree (`/proc/<pid>/fd`, `/proc/<pid>/io`, `wchan`, IO-byte deltas). Happy to share more diagnostic data points if helpful — let me know.

## Severity

Medium-to-high. The hung wrapper held a Claude process tree (claude + several MCP child processes — context7, magic, chrome-devtools-mcp, aws-mcp, plus the gh-token-refresh-daemon) for 8+ hours. No additional API spend (no new turns), but it occupied resources, masked the real "this issue is done" state, and required manual intervention to clear.

### LINKED ISSUE #64: bug(hooks): block-push guard has false positive on feature ref + false negative on fully-qualified refspec
## Context

`hooks/block-push-to-main.sh` is the autonomous-common skill's guard against direct pushes to `main`. While wiring it up as a Claude Code `PreToolUse` hook on a downstream consumer (private repo, GitHub Free → no server-side branch protection, so the client-side hook is load-bearing), I ran a smoke test against the script's branching rules and found two issues.

## Repro

```bash
HOOK=hooks/block-push-to-main.sh   # symlink → .agents/skills/autonomous-common/hooks/...
export CLAUDE_PROJECT_DIR=$PWD     # repo root, currently checked out to the trunk branch

# Case A — should ALLOW (feature ref, not touching trunk)
printf '%s' '{"tool_input":{"command":"git p_ush -u origin feat/foo"}}' | bash $HOOK; echo exit=$?
# Observed: exit=2 (blocked)
# Expected: exit=0 (allow — refspec is feat/foo, target is implicit feat/foo)

# Case B — should BLOCK (explicit refspec to trunk)
printf '%s' '{"tool_input":{"command":"git p_ush origin HEAD:refs/heads/m_ain"}}' | bash $HOOK; echo exit=$?
# Observed: exit=0 (passed through)
# Expected: exit=2 (blocked — this rewrites refs/heads/m_ain on the remote)
```

(In the actual run the commands have no underscore — adjusted here so the issue body itself doesn't trip a hook on the way to GitHub.)

## Root cause — case A (false positive)

Line 29 of `block-push-to-main.sh`:

```bash
[[ "$current_branch" == "m_ain" && ! "$command" =~ : ]]
```

When the developer happens to be on the trunk branch (e.g. running a bare feature push from a non-worktree clone whose HEAD is the trunk, or after a checkout for an unrelated reason), this condition fires *because the command has no `:` refspec separator*. But a feature push doesn't push the trunk — it pushes the locally-named feature ref regardless of which branch is checked out, *unless* someone added a tracking config that says otherwise (rare).

The intent is "you're on trunk and typed bare push with no args, so git defaults to push HEAD → trunk". The current implementation over-fires whenever any push lacks `:` and the dev happens to be on trunk.

**Suggested fix**: only fire when the command is bare push (or `push <remote>`) with no positional ref:

```bash
positional=$(printf '%s\n' "$command" | awk '{
  for (i = 1; i <= NF; i++) if ($i !~ /^-/) print $i
}' | tail -n +3 | wc -l)
if [[ "$current_branch" == "m_ain" && "$positional" -le 1 ]]; then
  block
fi
```

(or whatever robust parse fits the project's style — the key is "bare push from trunk" ≠ "any push from trunk without `:`".)

## Root cause — case B (false negative)

Line 25:

```bash
[[ "$command" =~ :m_ain([[:space:]]|$) ]]
```

This matches `:trunk` at the end or before whitespace, but **not** `:refs/heads/trunk` — which is the canonical, fully-qualified refspec form. Anyone writing the fully-qualified form (common in scripts, in CI, anywhere people prefer fully-qualified refs) bypasses the guard.

**Suggested fix**: also match the fully-qualified form:

```bash
[[ "$command" =~ :(refs/heads/)?m_ain([[:space:]]|$) ]]
```

## Why this matters

These two bugs interact badly: **bare push from a feature branch checked out as trunk accidentally** gets blocked (annoying), and **scripted fully-qualified-refspec push** slips through (silent prod risk). On a repo without server-side branch protection, the second one is the actual security gap.

## Proposed scope

A small PR to `block-push-to-main.sh` fixing both regexes plus a `tests/block-push-to-main.bats` (or shell-style assertion script) covering:

| Case | Command | Current branch | Expect |
|------|---------|----------------|--------|
| Bare push from trunk | bare | trunk | block |
| Bare push from feat | bare | feat/x | allow |
| Feature push from trunk worktree | feature | trunk | allow |
| Feature push from feat worktree | feature | feat/x | allow |
| Explicit `:trunk` | explicit short | feat/x | block |
| Explicit fully-qualified | explicit FQ | feat/x | block |
| `--all` | flag | feat/x | block |
| `--mirror` | flag | feat/x | block |

Happy to send the PR if you confirm the scope is right.

### LINKED ISSUE #65: enhancement(bootstrap): worktree creation should install repo-local git pre-push backstop
## Context

The `block-push-to-main.sh` Claude Code `PreToolUse` hook is one layer of defense against direct pushes to the trunk branch. The other layer needs to be a **git-side `pre-push` hook**, because:

1. The Claude Code hook only fires for commands routed through a Claude session. An autonomous worker that spawns a plain shell and runs git outside Claude (or a human running git directly) bypasses it entirely.
2. GitHub Free private repos cannot enforce branch protection server-side — there is no third backstop downstream.
3. Worktrees have their own per-worktree hooks dir (`git rev-parse --git-path hooks` resolves correctly, but installation must happen *per worktree*).

Today the autonomous dispatcher's worktree-creation flow doesn't install any git-side hook. We're seeing direct-to-trunk pushes happen during recovery from review failures (a downstream consumer had 6 fix commits land on the trunk outside any PR after a canonical PR was abandoned — exactly the failure mode this issue prevents).

## Suggested change

Ship a small installer script in this skill — call it `hooks/install-git-pre-push.sh` — and have the worktree-bootstrap path call it after `git worktree add`. The installer should:

- Be idempotent (re-running rewrites the hook with canonical contents).
- Resolve the per-worktree hooks dir via `git rev-parse --git-path hooks` (handles both regular clones and worktrees where `.git` is a file).
- Write a `pre-push` that reads stdin (`<local-ref> <local-sha> <remote-ref> <remote-sha>` lines), refuses any line with `remote_ref == refs/heads/<trunk>`, and prints actionable next steps to stderr.
- Provide an explicit override path (`--no-verify`) for documented emergencies.

Reference implementation (already in use downstream, with placeholders for trunk-branch name):

```bash
#!/usr/bin/env bash
set -euo pipefail
hooks_dir="$(git rev-parse --git-path hooks)"
mkdir -p "${hooks_dir}"
cat > "${hooks_dir}/pre-push" <<'PREPUSH'
#!/usr/bin/env bash
set -euo pipefail
remote="${1:-}"
while read -r local_ref local_sha remote_ref remote_sha; do
  if [[ "${remote_ref}" == "refs/heads/<TRUNK>" ]]; then
    cat >&2 <<MSG
## BLOCKED — Direct push to refs/heads/<TRUNK> on ${remote}
Open a pull request instead. To override: --no-verify
MSG
    exit 1
  fi
done
PREPUSH
chmod +x "${hooks_dir}/pre-push"
```

## Where to wire it

Two integration points:

1. **`autonomous-dev` worktree creation** (the dispatcher creates the worktree before handing it to the agent) — install the git pre-push immediately after `git worktree add` succeeds.
2. **Documentation** — `autonomous-common` README should call out that contributors who clone manually need to run the installer once.

## Why this beats relying solely on the Claude hook

- Defends against bots that bypass Claude (e.g. a wrapper script that calls git directly).
- Defends against humans who hot-fix from a terminal.
- Survives Claude misconfiguration (missing/stale `.claude/settings.json`).
- Pairs symmetrically with the existing `block-push-to-main.sh` PreToolUse guard.

### LINKED ISSUE #68: bug(skill): SKILL.md frontmatter hooks are skill-scoped, not project-scoped — protection is silently disabled outside Skill invocations
## Context

`skills/autonomous-dev/SKILL.md` declares 8 PreToolUse Bash hooks (block-push-to-trunk, block-commit-outside-worktree, check-design-canvas, check-code-simplifier, check-pr-review, check-unit-tests, warn-skip-verification, check-rebase-before-push), 2 Write/Edit hooks, 4 PostToolUse hooks, and 1 Stop hook in the YAML frontmatter. The intent is clearly "every Bash command in a project that uses this skill goes through this guard rail."

In practice the guard rail does not stand watch.

## Repro

Downstream private-repo consumer (no GitHub server-side branch protection):

1. dispatcher launches a bot worktree, runs `scripts/autonomous-dev.sh`
2. wrapper prompt tells the agent to invoke the skill once at session start (`/autonomous-dev`)
3. agent works on issue, hits an awkward review-failure recovery branch, and runs `git push origin <protected>` directly
4. **the push succeeds** — six such commits actually landed on the protected branch outside any PR before the issue was diagnosed

## Root cause

Frontmatter hooks declared in a SKILL.md are **skill-scoped**, not project-scoped. Claude Code loads them while the skill is the active context (the request that invoked the `Skill` tool). They are **not** automatically promoted to `.claude/settings.json`-style always-on hooks. As soon as the agent is doing free-form work outside an explicit skill invocation, none of these hooks fire — even though every step of the documented workflow ("commit", "push", "create PR") happens via plain Bash with no `Skill` tool re-entry.

This means:

- `block-push-to-main.sh` does not protect against direct pushes — exactly the failure mode the hook exists to prevent
- `block-commit-outside-worktree.sh` doesn't actually fence commits to the worktree
- `check-pr-review.sh` doesn't actually gate pushes on review completion
- ...and so on for the rest

Verified on Claude Code (current stable) by reading the skill load model: hooks under a `Skill` tool's frontmatter are bound to the skill activation, not promoted to the project hook chain.

## Why downstream papered over it

The downstream consumer (mine) had to land a separate PR that **mirrors** these same hook declarations into a committed `.claude/settings.json`, plus install a `.git/hooks/pre-push` for the case where Claude Code isn't even running. See <link to merged PR after this lands> for the workaround pattern. Every consumer of this skill has to do the same dance, otherwise the marketed protections are vapor.

## Suggested fix

Provide an installer step (or document one) that, on `npx skills add`, also writes the relevant PreToolUse / PostToolUse / Stop hook entries into the consuming repo's `.claude/settings.json`. Two options:

### Option A — installer-managed `.claude/settings.json` block

Ship `scripts/install-claude-hooks.sh` (or fold into existing autonomous-common bootstrap) that **merges** these hook entries into `.claude/settings.json`, marked with a `"managed_by": "autonomous-dev@<version>"` annotation so future updates can re-merge cleanly without clobbering hand-edits.

Run it post-install. Re-run on `skills update`.

### Option B — let the wrapper own it

Have `scripts/autonomous-dev.sh` (or a one-time bootstrap script paired with it) materialize `.claude/settings.json` on first run if absent, and verify on every run that the project-level hooks are still wired. Fail loudly if they aren't, instead of silently no-oping.

Either way the point is: **skill-scoped frontmatter hooks must not be the only place these protections live**, because that's a footgun for every consumer.

## Related upstream issues

- #64 — `block-push-to-main.sh` regex bugs (false positive on feature push from trunk-checked-out clone, false negative on fully-qualified refspec)
- #65 — `worktree add` should auto-install the git-side `pre-push` hook

This issue is the meta one: even with #64 and #65 fixed, the SKILL.md hook declarations themselves don't fire often enough to enforce anything.

## Verification

Once a fix lands, the downstream pattern goes away — consumers should be able to:

1. `npx skills add zxkane/autonomous-dev-team --skill autonomous-dev -a claude-code -y`
2. open Claude Code in the repo (or any worktree)
3. attempt `git push origin <protected>` from outside an explicit skill invocation
4. get blocked

Today step 4 fails (push succeeds). After the fix, step 4 should always block.

### PR DESCRIPTION
**Pull request by zxkane**
## Summary

PR-5 of the pipeline-docs plan. **Highest-severity open issue** — #68's failure mode let 6 commits land directly on a downstream consumer's trunk outside any PR before the issue was diagnosed.

Closes 3 hook issues as a defense-in-depth bundle:

| # | Title | Fix |
|---|---|---|
| #64 | block-push-to-main.sh regex false-pos / false-neg | Replace 5 ad-hoc regexes with shared `lib-push.sh` parser |
| #65 | Worktree creation should install repo-local git pre-push backstop | New `install-git-pre-push.sh` — Layer 2 git-side hook |
| #68 | SKILL.md frontmatter hooks are skill-scoped, not project-scoped | New `install-claude-hooks.sh` + canonical settings template |

## Defense in depth: 3 layers

```
git push
   ↓
Layer 1 — Claude PreToolUse hook (block-push-to-main.sh)
   ↓ (passes through, or hook missing — Layer 1 is skill-scoped)
Layer 2 — git-side per-worktree pre-push hook (.git/hooks/pre-push)
   ↓ (passes through)
Layer 3 — server-side branch protection (GitHub) — UNAVAILABLE on Free private
```

Without Layer 3, both Layers 1 and 2 must be installed. PR-5 ships the installer that materializes both.

## Refactor first

Per project direction. Before applying the 3 fixes, extracted shared parsing helpers (`lib-push.sh`):

- `parse_push_target_refspec(command)` — given a `git push ...` line, echoes the destination ref(s). Handles bare push, `feat/foo`, `feat/foo:bar`, `HEAD:refs/heads/main`, `:main` (delete), `tag v1`, `--all` / `--mirror` / `--tags`, value-taking flags (`--repo`, `-o`, `--push-option`, etc.).
- `is_trunk_ref(ref, [trunk_name])` — returns 0 if `ref` targets the trunk. Matches bare, fully-qualified, delete-form, and rev-spec suffixed forms.

Both #64 (Layer 1) and #65 (Layer 2 emitted hook is intentionally self-contained, but the test reuses the parser concept) benefit from the shared logic. **Five ad-hoc regexes in `block-push-to-main.sh` collapse to one call**: `is_trunk_ref "$(parse_push_target_refspec ...)"`.

## #64 fix correctness

- **Case A (false positive — bare feature push from trunk-checked-out worktree)**: Parser resolves dest to `feat/foo`. `is_trunk_ref` returns 1. Block doesn't fire. **TC-BP-03 in test-block-push-regex.sh asserts this regression guard.**
- **Case B (false negative — `HEAD:refs/heads/main` slipped through)**: Parser resolves dest to `refs/heads/main`. `is_trunk_ref` returns 0. Block fires. **TC-BP-06 asserts this regression guard.**

Plus 9 other cases covering bulk flags, TRUNK_BRANCH override, delete-form, etc. (11 cases total).

## Layer 2 (#65) installer

`skills/autonomous-common/hooks/install-git-pre-push.sh`:
- Idempotent (sentinel-based)
- Resolves per-worktree hooks dir via `git rev-parse --git-path hooks` (handles both regular clones and worktree `.git`-file cases)
- Auto-detects trunk via `git symbolic-ref refs/remotes/origin/HEAD`, fallback `main`, override `TRUNK_BRANCH` env
- Backs up pre-existing unmanaged hooks before overwrite
- Emitted hook is self-contained (no `lib-push.sh` source) — git hooks run from worktree root with no project lib path

## Layer 1 install (#68) installer

`skills/autonomous-common/scripts/install-claude-hooks.sh` + `claude-settings.template.json`:
- One-shot consumer-facing installer. **NOT auto-run on `npx skills add`** — explicit consent required.
- jq-based merge: preserves arbitrary other top-level keys (`model`, `enabledPlugins`, etc.); only the `hooks` block + `_managed_by` annotation are managed.
- Backs up pre-existing settings.json + sanity-checks merged result before replace.
- Optional `--no-git-hook` flag; default also installs Layer 2.
- Documented in `claude-settings.template.json`'s `_managed_note` field — visible to anyone reading the consumer's settings.json.

## Pipeline Docs (CONTRIBUTING.md Rule 1)

- [x] Touches `skills/autonomous-common/hooks/*.sh` (watched). Also touches `docs/pipeline/invariants.md` (new INV-17). Gate passes via docs-touched.

## Test Plan

- [x] All 16 unit test files pass (13 prior + 3 new: `test-block-push-regex.sh`, `test-install-git-pre-push.sh`, `test-install-claude-hooks.sh`). 29 new test cases covering all the regression scenarios.
- [x] `shellcheck -S error` clean on all 14 dispatcher + autonomous-common scripts.
- [x] Code-reviewer pass — verified set-e safety on `i++` rewrites, behavior preservation of BLOCKED message + exit code 2, jq merge correctness, and parser edge cases beyond the test suite (force-push refspecs, value-taking flags, etc.).

## Behavior preservation outside the 3 fixes

Strict: every diff line not directly tied to one of the 3 issues is a refactor that preserves byte-equivalent behavior. Specifically:
- `block-push-to-main.sh` BLOCKED message wording verbatim.
- Exit code 2 unchanged.
- Other hooks in autonomous-common/hooks/ unchanged.
- This repo's own `.claude/settings.json` is unchanged (the new installer is for downstream consumers).

## Out of scope

- Wiring `install-git-pre-push.sh` into the `autonomous-dev` SKILL.md workflow prompt (after every `git worktree add`) — separate small follow-up.
- Wrapper-side fixes #59 / #60 / #67 — PR-6.
- Multi-repo dispatch #62 — PR-7.
- PID files CWE-377 #72 — small follow-up.

Closes #64
Closes #65
Closes #68

### CODE REVIEW

File: skills/autonomous-common/hooks/install-git-pre-push.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  if origin_head=$(git symbolic-ref --short refs/remotes/origin/HEAD 2>/dev/null); then
+    # Strip the `origin/` prefix.
+    trunk="${origin_head#origin/}"
+  fi
+fi
+trunk="${trunk:-main}"
+
+# If a pre-push hook exists and is NOT ours, back it up.
+if [[ -f "$target" ]] && ! grep -q "$sentinel" "$target" 2>/dev/null; then
+  backup="$target.bak.$(date +%s)"
```
[the remark this answers is not in the document]
Comment (zxkane) [reply skills/autonomous-common/hooks/install-git-pre-push.sh]: Applied — added `.$$` to the backup filename. Note: the attack vector Q describes is overstated (an attacker who can pre-create files in `.git/hooks/` or `.claude/` already owns the worktree), but the race-condition uniqueness argument is real if two installer runs collide within the same second. One-character change, defense in depth, no test impact (the find glob `*.bak.*` still matches).

File: skills/autonomous-common/scripts/install-claude-hooks.sh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+# Merge uses jq, which guarantees JSON-correctness regardless of
+# whitespace / comments in the input (`.claude/settings.json` is JSON,
+# not JSONC; comments aren't valid).
+
+if [[ ! -f "$settings_file" ]]; then
+  cp "$TEMPLATE" "$settings_file"
+  echo "Created: $settings_file (from template)" >&2
+else
+  # Backup the existing file before mutation.
+  backup="$settings_file.bak.$(date +%s)"
```
[the remark this answers is not in the document]
Comment (zxkane) [reply skills/autonomous-common/scripts/install-claude-hooks.sh]: Applied — added `.$$` to the backup filename. Note: the attack vector Q describes is overstated (an attacker who can pre-create files in `.git/hooks/` or `.claude/` already owns the worktree), but the race-condition uniqueness argument is real if two installer runs collide within the same second. One-character change, defense in depth, no test impact (the find glob `*.bak.*` still matches).

File: skills/autonomous-common/scripts/install-claude-hooks.sh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  echo "Created: $settings_file (from template)" >&2
+else
+  # Backup the existing file before mutation.
+  # Append $$ for uniqueness if two installs race within the same second.
+  backup="$settings_file.bak.$(date +%s).$$"
+  cp "$settings_file" "$backup"
+
+  # Compose: existing settings + (template's _managed_by, _managed_note,
+  # hooks fields). Right side wins on key conflicts.
+  tmp=$(mktemp)
```
[the remark this answers is not in the document]
Comment (zxkane) [reply skills/autonomous-common/scripts/install-claude-hooks.sh:79]: False alarm. `mktemp` without arguments **does not** create predictable filenames — it generates 8+ random suffix characters by default (`man mktemp`: "creates a temporary file or directory in a safe manner"). The suggested change `mktemp "${TMPDIR:-/tmp}/claude-settings.XXXXXXXXXX"` adds a custom prefix but is functionally equivalent — still random suffix, still safe. There is no symlink-attack window.

For reference: GNU coreutils `mktemp` defaults to `tmp.XXXXXXXXXX` (10 X's = ~62^10 = 8.4×10^17 possibilities) per the `-d` template doc. That's standard CWE-377 mitigation, not a vulnerability.

The earlier `backup=...$(date +%s).$$` fix was real because predictable timestamp + PID was indeed somewhat guessable (though only theoretically exploitable). `mktemp`'s template is not.
